draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,431 @@
+
+  parse(arguments)
+end
+```
+
+Ok, nice and easy. First we'll set the configuration to our defaults, then we're parsing the cli arguments and overwrite our configuration based on this.
+
+What does the default configuration look like?
+For that you'll have to check out the top level library file, so lib/mutant.rb:
+
+
+```Ruby
+module Mutant
+  # Reopen class to initialize constant to avoid dep circle
+  class Config
+    DEFAULT = new(
+      debug:             false,
+      expected_coverage: Rational(1),
+      # snip
+      fail_fast:         false,
+      includes:          EMPTY_ARRAY,
+      # snip
+      kernel:            Kernel,
+      load_path:         $LOAD_PATH,
+      matcher:           Matcher::Config::DEFAULT,
+      open3:             Open3,
+      pathname:          Pathname,
+      requires:          EMPTY_ARRAY,
+      reporter:          Reporter::CLI.build($stdout),
+      zombie:            false
+    )
+  end # Config
+end # Mutant
+```
+
+I did remove a lot of the config object above to keep it readable. I just want to highlight a couple of lines I find interesting:
+
+```Ruby
+expected_coverage: Rational(1)
+```
+
+This means that the expected coverage is 100%. Yep, 100%.
+
+```Ruby
+reporter:          Reporter::CLI.build($stdout)
+```
+
+The default reporter. Keep that in mind, we'll come back to this later.
+
+```Ruby
+load_path:         $LOAD_PATH,
+requires:          EMPTY_ARRAY,
+pathname:          Pathname,
+```
+
+What I like here is how configurable `Mutant`. You can even inject another $LOAD_PATH, additional requires or another Pathname'esque library.
+
+**The big picture for now:**
+
+TODO: proper picture
+
+bin -> cli -> config
+
+### Environment
+
+So where were we?
+Keep in mind that we are still talking about this:
+
+```Ruby
+Runner.call(
+  Env::Bootstrap.call(
+    call(arguments) # -> This part we just discussed.
+  )
+).success?
+```
+
+Now that we know what 
+
+```Ruby
+call(arguments) # -> This part we just discussed.
+```
+
+does, let's look at:
+
+```Ruby
+Env::Bootstrap.call(
+  call(arguments) # -> This part we just discussed.
+)
+```
+
+Again, this is `Procto` at work.
+
+```Ruby
+Env::Bootstrap.call configuration
+```
+
+means
+
+```Ruby
+ # "some_method" is what we configure when including Procto
+Env::Bootstrap.new(configuration).some_method
+```
+
+So let's check out the class declaration of Env::Bootstrap with a focus on the initializer:
+
+```Ruby
+module Mutant
+  class Env
+    class Bootstrap
+      include Procto.call(:env) # -> That's the "some_method" from above
+
+      def initialize(*)
+        super
+        @parser = Parser.new
+        infect
+        initialize_matchable_scopes
+      end
+    end
+  end
+end
+```
+
+Step by step:
+
+* `super`: TODO: Explain Concord?
+* `Parser.new`: `Mutant` uses the awesome `[parser](https://github.com/whitequark/parser)` gem for producing ASTs from source code and has its own thin wrapper around
+* `infect`: TODO
```

Comment:
Infection is the process where mutant processes includes and requires, and "infects" the ruby VM with the "subjects under tests", its the point where the software that is under mutation test gets loaded.

I called it "infection" as from now on all the ugly stuff happens that sadly "typical" ruby projects do, monkeypatching in particular.